enterprise/server/util/vfs_server/vfs_server.go

```diff
@@ -411,63 +418,27 @@ func (p *Server) createLayout(ctx context.Context, inputTree *repb.Tree, digestF
 		return nil
 	}
 
-	err = walkDir(inputTree.Root, rootNode)
+	err = walkDir(inputTree.Root, p.root)
 	if err != nil {
-		return nil, err
+		return err
 	}
 
 	log.CtxDebugf(ctx, "VFS contains %d directories, %d files and %d symlinks", numDirs, numFiles, numSymlinks)
 
-	return rootNode, nil
-}
-
-func (p *Server) mkdirAll(root *fsNode, path string) {
-	node := root
-	for _, dir := range strings.Split(path, string(os.PathSeparator)) {
-		if subNode, ok := node.children[dir]; ok {
-			node = subNode
-			continue
-		}
-
-		node.mu.Lock()
-		if node.children == nil {
-			node.children = make(map[string]*fsNode)
-		}
-		subNode := newDirNode(node, dir)
-		p.addNode(subNode)
-		node.children[dir] = subNode
-		node.mu.Unlock()
-		node = subNode
-	}
+	return nil
 }
 
 // Prepare is used to inform the VFS server about files that can be lazily loaded on the first open attempt.
 func (p *Server) Prepare(ctx context.Context, layout *container.FileSystemLayout) error {
-	rootNode, err := p.createLayout(ctx, layout.Inputs, layout.DigestFunction)
+	err := p.updateLayout(ctx, layout.Inputs, layout.DigestFunction)
```

Comment:
The workspace code is now creating the output dirs which means the special output directory handling here is no longer needed. 

A side-effect of this is that the output directories are created before `Prepare` is called which means instead of swapping out the entire tree, we need to update the existing node tree to merge the information the CAS input tree.